### PR TITLE
feat: add authentication for RPC service

### PR DESCRIPTION
**Pull request by lauti7**
This PR: 
- implements https://github.com/decentraland/adr/pull/229 even though it could change

### CODE REVIEW

File: crates/server/src/rpc/mod.rs [hunk 1]
```diff
@@ -121,3 +146,60 @@ async fn handle_rejection(err: Rejection) -> Result<impl Reply, std::convert::In
...
+                if let Ok(address) = authenticator
+                    .verify_signature(&auth_chain, &message_to_be_firmed)
+                    .await
+                {
+                    let address = address.to_owned();
+                    match ws_write.reunite(ws_read) {
+                        Ok(ws) => Ok((ws, address)),
+                        Err(err) => Err(AuthenticationErrors::UnexpectedError(Box::new(err))),
+                    }
+                } else if let Err(err) = ws_write.close().await {
```
Comment (guidota) [inline crates/server/src/rpc/mod.rs:188]: What do you say if we close the websocket in the upgrade handler and return the error first?

File: crates/server/src/rpc/mod.rs [hunk 2]
```diff
@@ -121,3 +146,60 @@ async fn handle_rejection(err: Rejection) -> Result<impl Reply, std::convert::In
...
+        .send(Message::text(&message_to_be_firmed))
+        .await
+        .is_err()
+    {
+        return Err(AuthenticationErrors::FailedToSendChallenge);
+    }
+
+    match tokio::time::timeout(Duration::from_secs(30), ws_read.next()).await {
+        Ok(client_response) => {
+            let response = client_response.unwrap().unwrap();
```
Comment (guidota) [inline crates/server/src/rpc/mod.rs:176]: Shouldn't we handle the possible errors here?

File: crates/server/src/rpc/mod.rs [hunk 3]
```diff
@@ -121,3 +146,60 @@ async fn handle_rejection(err: Rejection) -> Result<impl Reply, std::convert::In
         ))
     }
 }
+
+pub enum AuthenticationErrors {
+    FailedToSendChallenge,
+    WrongSignature,
+    Timeout,
+    NotTextMessage,
+    UnexpectedError(Box<dyn std::error::Error + Send + Sync>),
```
Comment (guidota) [inline crates/server/src/rpc/mod.rs:155]: If we move the socket close to the handler, then we don't need this one.

File: crates/server/src/rpc/mod.rs [hunk 4]
```diff
@@ -121,3 +146,60 @@ async fn handle_rejection(err: Rejection) -> Result<impl Reply, std::convert::In
         ))
     }
 }
+
+pub enum AuthenticationErrors {
```
Comment (guidota) [inline crates/server/src/rpc/mod.rs:150]:
```suggestion
pub enum AuthenticationError {
```

**Review by guidota — CHANGES_REQUESTED** [review]
I think it would be great to move the authentication stuff to its own module.

File: crates/server/src/rpc/mod.rs [hunk 5]
```diff
@@ -121,3 +146,60 @@ async fn handle_rejection(err: Rejection) -> Result<impl Reply, std::convert::In
...
+        .is_err()
+    {
+        return Err(AuthenticationErrors::FailedToSendChallenge);
+    }
+
+    match tokio::time::timeout(Duration::from_secs(30), ws_read.next()).await {
+        Ok(client_response) => {
+            let response = client_response.unwrap().unwrap();
+            if let Ok(auth_chain) = response.to_str() {
+                let auth_chain = AuthChain::from_json(auth_chain).unwrap();
```
Comment (guidota) [inline crates/server/src/rpc/mod.rs:178]: Shouldn't we handle possible errors here?

File: crates/server/src/rpc/mod.rs (hunk 1 above)
Comment (lauti7) [reply crates/server/src/rpc/mod.rs:188]: Okay but the error would be `Err((Websocket,AuthenticationErrors))` because we have to consume the ws

**Review by guidota — APPROVED** [review]

**Pull request merged**